src/namespace.rs

```diff
@@ -0,0 +1,38 @@
+use std::borrow::Cow;
+
+/// Types that can be used as top-level storage keys for storage items and collections.
+pub trait Namespace {
+    fn namespace(self) -> Ns;
+}
```

Comment:
Can this be implemented if Ns is private? If not, let's document document that.

What about calling this trait `IntoNs`?

Reply:
Doesn't `IntoNs` sound somewhat cryptic from the library consumer PoV? I'm trying to hide the `Ns` type here, mind you.

Reply:
Good point.

But then again you have a type X: Namespace. Why would you call namespace to turn it into a Ns if it is already a namespace?

What's the motivation for hiding by the way? The inner type of Ns is private anyways so we can change it any time, even if public.

Reply:
> But then again you have a type X: Namespace. Why would you call namespace to turn it into a Ns if it is already a namespace?

Because I need to transform it into the internal representation of the namespace. It's still the same thing, in some sense. Would it help to rename `Ns` to something like `NamespaceInner`?

> What's the motivation for hiding by the way?

What's the benefit of exposing it? The goal was never to add another type consumers would have to worry about, but to let them use either `&'static str` or `String` in constructors, and sometimes for internal code to call a constructor reusing an `Ns` object. I don't want to add unnecessary cognitive load.

Reply:
> What's the benefit of exposing it?

You only need one type instead of two.

But what about using `Into<String>` as the constructor function argument and don't even make the trait public? This avoid the need to expose the type alltogether and creates a pretty simple interface.

Reply:
> But what about using Into<String> as the constructor function argument and don't even make the trait public?

With `Into<String>` there's no way to handle the case we get a reference other than by converting to a `String`. Some constructors take multiple storage keys and I figured we'd want to allow people to mix and match `&'static str`s and `String`s.

I think `Into<Cow<'static, str>>` is more likely to work well here. It makes it a bit harder to deal with [places where we want to reuse `Ns`](https://github.com/CosmWasm/cw-storage-plus/pull/68/files#diff-8784c8bce77d653b87e91a641306152e2f91f2ddd369c9eedf7b6a0452380186R53-R54). I could add a `Clone` trait bound or, if we're sure we don't want to allow binary keys in the future, make `Ns` store a `Cow<'static, str>` rather than `Cow<'static, [u8]>`.

Reply:
Right.

In this case my preference would be to just have the `Namespace` struct public and use `Into<Namespace>` for `::new_dyn`. The implementation (currently Cow) is still private in that case. This might export slightly more than needed but reduces the number of types and relationships, leading to less cognitive load. In examples `Namespace` may never be used as you can just have strings in there.

Reply:
I don't know if this really is less cognitive load given this approach caused some confusion earlier in the PR ;) But I'll do it.